crates/red_knot_python_semantic/resources/mdtest/protocols.md

```diff
@@ -1,9 +1,1015 @@
+```
+
+Attribute members on protocol classes are allowed to have default values, just like instance
+attributes on other classes. Similar to nominal classes, attributes
+with defaults can be accessed on the class object itself and any explicit subclasses of the
+protocol class. It cannot be assumed to exist on the meta-type of any arbitrary inhabitant of the
+protocol type, however; an implicit subtype of the protocol will not necessarily have a default value
+for the instance attribute provided in its class body:
+
+```py
+class HasXWithDefault(Protocol):
+    x: int = 42
+
+reveal_type(HasXWithDefault.x)  # revealed: int
+
+class ExplicitSubclass(HasXWithDefault): ...
+
+reveal_type(ExplicitSubclass.x)  # revealed: int
+
+def f(arg: HasXWithDefault):
+    # TODO: should emit `[unresolved-reference]` and reveal `Unknown`
+    reveal_type(type(arg).x)  # revealed: int
+```
+
+Attribute members are allowed to have assignments in methods on the protocol class, just like
+non-protocol classes. Unlike other classes, however, *implicit* instance attributes -- those that
+are not declared in the class body -- are not allowed:
+
+```py
+class Foo(Protocol):
+    x: int
+    y: str
+
+    def __init__(self) -> None:
+        self.x = 42  # fine
+        self.a = 56  # error
+
+    def non_init_method(self) -> None:
+        self.y = 64  # fine
+        self.b = 72  # error
+```
+
+If a protocol has 0 members, then all other types are assignable to it, and all fully static types are subtypes of it:
+
+```py
+from typing import Protocol
+
+class UniversalSet(Protocol): ...
+
+static_assert(is_assignable_to(object, UniversalSet))  # error: [static-assert-error]
+static_assert(is_subtype_of(object, UniversalSet))  # error: [static-assert-error]
+```
+
+Which means that `UniversalSet` here is in fact an equivalent type to `object`:
+
+```py
+from knot_extensions import is_equivalent_to
+
+static_assert(is_equivalent_to(UniversalSet, object))  # error: [static-assert-error]
+```
+
+`object` is a subtype of certain other protocols too. Since all fully static types (whether nominal or structural)
+are subtypes of `object`, these protocols are also subtypes of `object`; and this means that these protocols
+are also equivalent to `UniversalSet` and `object`:
+
+```py
+class SupportsStr(Protocol):
+    def __str__(self) -> str: ...
+
+static_assert(is_equivalent_to(SupportsStr, UniversalSet))  # error: [static-assert-error]
+static_assert(is_equivalent_to(SupportsStr, object))  # error: [static-assert-error]
+
+class SupportsClass(Protocol):
+    __class__: type
+
+static_assert(is_equivalent_to(SupportsClass, UniversalSet))  # error: [static-assert-error]
+static_assert(is_equivalent_to(SupportsClass, SupportsStr))  # error: [static-assert-error]
+static_assert(is_equivalent_to(SupportsClass, object))  # error: [static-assert-error]
+```
+
+A protocol can never be assignable to (or a subtype of) any nominal type other than `object`, however:
```

Comment:
The code block below seems to be missing any demonstration of this, it only shows that they are assignable/subtype of `object`.

Reply:
oops, thanks